Realm/ObjectStore/impl/async_query.cpp

```diff
@@ -0,0 +1,265 @@
+            return;
+        }
+    }
+
+    m_tv = m_query->find_all();
+    if (m_sort) {
+        m_tv.sort(m_sort.columnIndices, m_sort.ascending);
+    }
+}
+
+void AsyncQuery::prepare_handover()
+{
+    m_sg_version = m_sg->get_version_of_current_transaction();
+
+    if (!m_tv.is_attached()) {
+        return;
+    }
+
+    REALM_ASSERT(m_tv.is_in_sync());
+
+    m_initial_run_complete = true;
+    m_handed_over_table_version = m_tv.outside_version();
+    m_tv_handover = m_sg->export_for_handover(m_tv, MutableSourcePayload::Move);
+    m_tv = TableView();
+}
+
+bool AsyncQuery::deliver(SharedGroup& sg, std::exception_ptr err)
+{
+    if (!is_for_current_thread()) {
+        return false;
+    }
+
+    std::lock_guard<std::mutex> target_lock(m_target_mutex);
+
+    // Target results being null here indicates that it was destroyed while we
+    // were in the process of advancing the Realm version and preparing for
+    // delivery, i.e. it was destroyed from the "wrong" thread
+    if (!m_target_results) {
+        return false;
+    }
+
+    // We can get called before the query has actually had the chance to run if
+    // we're added immediately before a different set of async results are
+    // delivered
+    if (!m_initial_run_complete && !err) {
+        return false;
+    }
+
+    if (err) {
+        m_error = err;
+        return m_have_callbacks;
+    }
+
+    REALM_ASSERT(!m_query_handover);
+
+    auto realm_sg_version = Realm::Internal::get_shared_group(*m_realm).get_version_of_current_transaction();
+    if (m_sg_version != realm_sg_version) {
+        // Realm version can be newer if a commit was made on our thread or the
+        // user manually called refresh(), or older if a commit was made on a
+        // different thread and we ran *really* fast in between the check for
+        // if the shared group has changed and when we pick up async results
+        return false;
+    }
+
+    if (m_tv_handover) {
+        m_tv_handover->version = m_sg_version;
+        Results::AsyncFriend::set_table_view(*m_target_results,
+                                             std::move(*sg.import_from_handover(std::move(m_tv_handover))));
+        m_delievered_table_version = m_handed_over_table_version;
+
+    }
+    REALM_ASSERT(!m_tv_handover);
+    return m_have_callbacks;
+}
+
+void AsyncQuery::call_callbacks()
+{
+    REALM_ASSERT(is_for_current_thread());
+
+    while (auto fn = next_callback()) {
+        fn(m_error);
+    }
+
+    if (m_error) {
+        // Remove all the callbacks as we never need to call anything ever again
+        // after delivering an error
+        std::lock_guard<std::mutex> callback_lock(m_callback_mutex);
+        m_callbacks.clear();
+    }
+}
+
+std::function<void (std::exception_ptr)> AsyncQuery::next_callback()
+{
+    std::lock_guard<std::mutex> callback_lock(m_callback_mutex);
+    for (++m_callback_index; m_callback_index < m_callbacks.size(); ++m_callback_index) {
```

Comment:
If I'm reading this correctly, this is using unsigned integer overflow to get `m_callback_index` from its default value of `npos` into the range of indices that are valid for `m_callbacks`. Is there a reason this approach is used rather than having `call_callbacks` set `m_callback_index` to zero prior to making the first call to `next_callback`?